### PR TITLE
Add active filter

### LINKED ISSUE #1312: Add query filter for conversation status
# Problem

- We need to change `ConversationQuery.status_filter` to pattern match on `%{"active" => true}` and rename it `active_filter` (likely requiring a change to the client)
- We need to then add a `status_filter` that filters on the open/closed status

### PR DESCRIPTION
**Pull request by snewcomer**
Fixes #1312

### CODE REVIEW

File: lib/code_corps/messages/conversation_query.ex [hunk 1]
```diff
@@ -53,5 +53,16 @@ defmodule CodeCorps.Messages.ConversationQuery do
...
+
+  @doc ~S"""
+  Narrows down a `CodeCorps.Conversation` query to return only those records
+  considered to have a specific status.
+  """
+  @spec status_filter(Queryable.t, map) :: Queryable.t
+  def status_filter(queryable, %{"status" => status}) do
+    queryable
+    |> where([c], c.status == ^status)
+  end
```
Comment (snewcomer) [inline lib/code_corps/messages/conversation_query.ex:65]: This still needs a test

File: test/lib/code_corps/messages/messages_test.exs [hunk 2]
```diff
@@ -180,23 +180,23 @@ defmodule CodeCorps.MessagesTest do
...
       refute conversation_started_by_admin_without_reply.id in result_ids
       assert conversation_started_by_admin_with_reply.id in result_ids
       assert conversation_started_by_user_without_reply.id in result_ids
       assert conversation_started_by_user_with_reply.id in result_ids
 
-      result_ids =
-        Conversation
-        |> Messages.list_conversations(%{"status" => "any"})
-        |> get_and_sort_ids()
+      # result_ids =
```
Comment (snewcomer) [inline test/lib/code_corps/messages/messages_test.exs]: Should the default be `Repo.all()`....do we want to get all conversations?

File: lib/code_corps/messages/conversation_query.ex [hunk 3]
```diff
@@ -41,8 +41,8 @@ defmodule CodeCorps.Messages.ConversationQuery do
   - belong to a `CodeCorps.Message` initiated by admin, with at least a single
     reply in the form of a `CodeCorps.ConversationPart`
   """
-  @spec status_filter(Queryable.t, map) :: Queryable.t
-  def status_filter(queryable, %{"status" => "active"}) do
+  @spec active_filter(Queryable.t, map) :: Queryable.t
+  def active_filter(queryable, %{"active" => true}) do
```
Comment (snewcomer) [inline lib/code_corps/messages/conversation_query.ex:45]: @joshsmith is this how active filter should work?  What specifically filters out inactive records?

File: lib/code_corps/messages/conversation_query.ex [hunk 4]
```diff
@@ -53,5 +53,16 @@ defmodule CodeCorps.Messages.ConversationQuery do
     |> having([_c, m, _cp], m.initiated_by == "user")
     |> or_having([c, m, cp], m.initiated_by == "admin" and count(cp.id) > 0)
```
Comment (joshsmith) [inline lib/code_corps/messages/conversation_query.ex:54]: In response to your question above, this is the query that does it (specifically the `having` and `or_having`), and the documentation above expands on that.

File: lib/code_corps/messages/conversation_query.ex [hunk 5]
```diff
@@ -41,8 +41,8 @@ defmodule CodeCorps.Messages.ConversationQuery do
   - belong to a `CodeCorps.Message` initiated by admin, with at least a single
```
Comment (joshsmith) [inline lib/code_corps/messages/conversation_query.ex]: We need to update this documentation.

> Filters `CodeCorps.Conversation` record queries to return only those considered to be active.
>
> Active conversations belong either:

And then update the bullets to remove "belong".

File: lib/code_corps/messages/conversation_query.ex [hunk 6]
```diff
@@ -53,5 +53,16 @@ defmodule CodeCorps.Messages.ConversationQuery do
     |> having([_c, m, _cp], m.initiated_by == "user")
     |> or_having([c, m, cp], m.initiated_by == "admin" and count(cp.id) > 0)
   end
-  def status_filter(query, %{}), do: query
+  def active_filter(query, %{}), do: query
+
+  @doc ~S"""
+  Narrows down a `CodeCorps.Conversation` query to return only those records
+  considered to have a specific status.
```
Comment (joshsmith) [inline lib/code_corps/messages/conversation_query.ex]: Would change this to:

> Filters `CodeCorps.Conversation` record queries by their status.

File: test/lib/code_corps/messages/messages_test.exs (hunk 2 above)
Comment (joshsmith) [reply test/lib/code_corps/messages/messages_test.exs]: I think the test on L128 still stands: `returns all records by default`.

In this case we just want to test filtering by that status.

Comment (joshsmith) [reply test/lib/code_corps/messages/messages_test.exs]: And to expand on this, the client really should be responsible for passing the "defaults" in cases where it's not about authorization to see records.

**Review by joshsmith — COMMENTED** [review]
A few things I'd recommend changing, along with some answers to your questions.

**Review by joshsmith — APPROVED** [review]

**Pull request merged**